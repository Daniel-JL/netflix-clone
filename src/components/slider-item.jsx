/* eslint-disable import/prefer-default-export */
import React, { useState, useRef, useEffect } from 'react';
import {
  NavLink,
} from 'react-router-dom';
import styled from 'styled-components';

const ISliderItemProps = {
  mediaName: 'name',
  imageID: 1,
  maturityRating: 'rating',
  numOfSeasons: 1,

};

const ItemContainer = styled.div`
  height: 139px;
  width: 250px;
  border: 1px solid black;
`;

const SliderItemImage = styled.img`
  width: 100%;
  height: 100%;
`;

export function SliderItem(props) {
  const [imgLoaded, setImgLoaded] = useState(false);
  const [imgLoadingErr, setImgLoadingErr] = useState(false);
  const [imgLoadedSuccess, setImgLoadedSuccess] = useState(false);
  const posterPath = useRef('');

  const fetchUrlData = async () => fetch(`https://api.themoviedb.org/3/${props.mediaType}/${props.mediaId}?api_key=${process.env.REACT_APP_MOVIE_DB_API_KEY}`)
    .then((response) => response.json())
    .then((data) => {
      // console.log(data.backdrop_path);
      posterPath.current = `http://image.tmdb.org/t/p/original${data.backdrop_path}`;
      setImgLoaded(true);
    })
    .catch((error) => {
      console.error('Error', error);
    });

  useEffect(() => {
    if (!imgLoaded) {
      fetchUrlData();
    }
  });

  return (
    <ItemContainer>
      {imgLoaded && !imgLoadingErr
      && (
      <SliderItemImage
        alt="Slider image"
        src={posterPath.current}
        onError={() => setImgLoadingErr(true)}
        onLoad={() => setImgLoadedSuccess(true)}
      />
      )}
      {imgLoadedSuccess &&
      <div id="imgSuccess" data-testid="imgSuccess" />
      }
    </ItemContainer>
  );
}
