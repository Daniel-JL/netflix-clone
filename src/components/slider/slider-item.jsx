/* eslint-disable import/prefer-default-export */
import React, { useRef, useState } from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';
import styled from 'styled-components';
import { Transition } from 'react-transition-group';
import {
  RoundPlayButton,
  RoundEpsAndInfoButton,
} from '../buttons';
import { Modal } from '../modal';

const ItemContainer = styled.div`

  position: relative;
  width: ${({ itemWidth }) => itemWidth}px;
  border: 1px solid black;
  z-index: 2;
`;

const SliderItemImage = styled.img`
  width: 100%;
  height: 100%;
`;

const ItemDetails = styled.div`
  
`;

const ModalItem = styled.div`
  position: absolute;
  display: flex;
  flex-direction: column;
  background-color: rgb(40,40,40);
  color: white;
  
  left: ${({ coordsLeft }) => coordsLeft}px;
  top: ${({ coordsTop }) => coordsTop}px;
  width: ${({ itemWidth }) => itemWidth}px;
`;

const ImgTitleContainer = styled.div`
  position: relative;
`;

const MediaTitle = styled.div`
  position: absolute;
  color: white;
  text-shadow: -1px -1px 0 #000, 1px -1px 0 #000, -1px 1px 0 #000, 1px 1px 0 #000;
  bottom: 10%;
  left: 2%;
  font-size: 1.3vw;
`;

export function SliderItem({
  mediaId,
  mediaType,
  sliderItemData,
  imgLoadedSuccess,
  imgLoadingErr,
  dataLoaded,
  itemHoverActive,
  itemHoverTransition,
  handleMouseOver,
  handleMouseOut,
  handleModalDismount,
  setImgLoadingErr,
  setImgLoadSuccess,
  handleEpsAndInfoButtonClick,
  itemDimensions,
}) {
  const divRef = useRef();
  const location = useLocation();
  const [itemContainerRef, setItemContainerRef] = useState();

  const duration = 200;

  if (itemHoverActive && divRef.current && divRef.current.matches(':hover') === false) {
    handleMouseOut();
  }

  const defaultStyle = {
    transition: `all ${duration}ms ease-in-out`,
    opacity: 1,
    position: 'absolute',
    display: 'flex',
    flexDirection: 'column',
    left: '0px',
    top: '0px',
    width: `${itemDimensions.width}px`,
  };

  const transitionStyles = {
    entering: {},
    entered: {
      opacity: 1,
      transform: 'scale(1.5)',
    },
    exiting: {},
    exited: {transform:'scale(0.5)'},
  };

  return (
    <ItemContainer
      ref={setItemContainerRef}
      onMouseEnter={() => handleMouseOver(itemContainerRef)}
      onMouseLeave={() => handleMouseOut()}
      active={itemHoverActive}
    >

      {dataLoaded && !imgLoadingErr
      && (
        <ImgTitleContainer>
          <MediaTitle>
            {sliderItemData.mediaTitle}
          </MediaTitle>
          <SliderItemImage
            alt="Slider image"
            src={sliderItemData.posterPath}
            onError={() => setImgLoadingErr()}
            onLoad={() => setImgLoadSuccess()}
          />
        </ImgTitleContainer>
      )}

      {/* {imgLoadedSuccess
      && <div id="imgSuccess" data-testid="imgSuccess" />} */}

      {itemHoverActive
        && (
        <Modal
          id="slider-item-modal"
          height={itemDimensions.height}
          width={itemDimensions.width}
          left={itemDimensions.left}
          top={itemDimensions.top}
          bottom={itemDimensions.bottom}
          right={itemDimensions.right}
        >
          <Transition
            appear
            in={itemHoverTransition}
            timeout={duration}
            onExited={() => handleModalDismount()}
          >
            {(state) => (
              <ModalItem
                style={{
                  ...defaultStyle,
                  ...transitionStyles[state],
                }}
                id="modal-item"
                coordsLeft={itemDimensions.left}
                coordsTop={itemDimensions.top}
                itemWidth={itemDimensions.width}
                onMouseLeave={() => handleMouseOut()}
                ref={divRef}
              >
                <ImgTitleContainer>
                  <MediaTitle>
                    {sliderItemData.mediaTitle}
                  </MediaTitle>
                  <SliderItemImage
                    alt="Slider image"
                    src={sliderItemData.posterPath}
                    onError={() => setImgLoadingErr()}
                    onLoad={() => setImgLoadSuccess()}
                  />
                </ImgTitleContainer>
                <ItemDetails>
                  <div id="buttons">
                    <Link
                      key={1}
                      to="/watch"
                    >
                      <RoundPlayButton>
                        |>
                      </RoundPlayButton>
                    </Link>
                    <Link
                      key={1}
                      to={{
                        pathname: `${location.pathname}/epsinfobox/${mediaType}/${mediaId}`,
                        state: { background: location },
                      }}
                    >
                      <RoundEpsAndInfoButton onClick={handleEpsAndInfoButtonClick}>
                        v
                      </RoundEpsAndInfoButton>
                    </Link>
                  </div>
                  <div id="media-info">
                    {sliderItemData.ageRating}
                    {' '}
                    {sliderItemData.runtimeOrNumberOfSeasons > 1 && `${sliderItemData.runtimeOrNumberOfSeasons.length} Seasons`}
                    {sliderItemData.runtimeOrNumberOfSeasons === 1 && `${sliderItemData.runtimeOrNumberOfSeasons.length} Season`}
                  </div>
                  <div id="genres">
                    {sliderItemData.genres[0]}
                    {/* {'  '}
                    {'  '}
                    {genres[1]} */}
                  </div>
                </ItemDetails>
              </ModalItem>
            )}
          </Transition>
        </Modal>
        )}
    </ItemContainer>
  );
}
