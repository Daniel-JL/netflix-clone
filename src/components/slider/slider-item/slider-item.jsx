import React, { useRef, useState } from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';
import styled from 'styled-components';
import { Transition } from 'react-transition-group';
import {
  RoundPlayButton,
  RoundEpsAndInfoButton,
} from '../../common/buttons/buttons';
import Modal from '../../main-page/modal/modal';

const ItemContainer = styled.div`
  position: relative;
  width: ${({ itemWidth }) => itemWidth}px;
  border: 1px solid black;
  border-radius: 0.2vw;
  z-index: 2;
`;

const SliderItemImage = styled.img`
  width: 100%;
  height: 100%;
  border-radius: 0.2vw;

  ${({ mouseOver }) => mouseOver
    && `
      border-top-right-radius: 0.2vw;
      border-top-left-radius: 0.2vw;
      border-bottom-right-radius: 0px;
      border-bottom-left-radius: 0px;
    `
} 
`;

const ItemDetails = styled.div`
  
`;

const ModalItem = styled.div`
  position: absolute;
  display: flex;
  flex-direction: column;
  background-color: rgb(33, 33, 33);
  color: white;
  border-radius: 0.2vw;
  
  left: ${({ coordsLeft }) => coordsLeft}px;
  top: ${({ coordsTop }) => coordsTop}px;
  width: ${({ itemWidth }) => itemWidth}px;
`;

const ButtonsContainer = styled.div`
  display: flex;
  justify-content: space-between;
  padding-left: 2%;
  width: 4.5vw;
`;

const InfoContainer = styled.div`
  display: flex;
  padding-left: 2%;
  padding-top: 2%;
  font-size: 13px;
`;

const AgeRatingContainer = styled.div`
  border: 0.1px solid white;
  padding: 0.5px;
  padding-left: 2px;
  padding-right: 2px;
`;

const GenreContainer = styled.div`
  display: flex;
  padding: 2%;
  font-size: 13px;
`;

const ImgTitleContainer = styled.div`
  position: relative;
`;

const MediaTitle = styled.div`
  position: absolute;
  color: white;
  text-shadow: -1px -1px 0 #000, 1px -1px 0 #000, -1px 1px 0 #000, 1px 1px 0 #000;
  bottom: 10%;
  left: 2%;
  font-size: 1.3vw;
`;

const SliderItem = ({
  mediaId,
  mediaType,
  sliderItemData,
  imgLoadedSuccess,
  imgLoadingErr,
  dataLoaded,
  itemHoverActive,
  itemHoverTransition,
  handleMouseOver,
  handleMouseOut,
  handleModalDismount,
  setImgLoadingErr,
  setImgLoadSuccess,
  handleEpsAndInfoButtonClick,
  itemDimensions,
}) => {
  const divRef = useRef();
  const location = useLocation();
  const [itemContainerRef, setItemContainerRef] = useState();

  const duration = 200;

  if (itemHoverActive && divRef.current && divRef.current.matches(':hover') === false) {
    handleMouseOut();
    console.log(sliderItemData.runtimeOrNumberOfSeasons);
  }

  const defaultStyle = {
    transition: `all ${duration}ms ease-in-out`,
    opacity: 1,
    position: 'absolute',
    display: 'flex',
    flexDirection: 'column',
    left: '0px',
    top: '0px',
    width: `${itemDimensions.width}px`,
  };

  const transitionStyles = {
    entering: {},
    entered: {
      opacity: 1,
      transform: 'scale(1.5)',
    },
    exiting: {},
    exited: { transform: 'scale(0.5)' },
  };

  return (
    <ItemContainer
      ref={setItemContainerRef}
      onMouseEnter={() => handleMouseOver(itemContainerRef)}
      onMouseLeave={() => handleMouseOut()}
      active={itemHoverActive}
    >
      {/* If img is loaded correctly then it is mounted here */}
      {dataLoaded && !imgLoadingErr
      && (
        <ImgTitleContainer>
          <MediaTitle>
            {sliderItemData.mediaTitle}
          </MediaTitle>
          <SliderItemImage
            alt="Slider image"
            src={sliderItemData.posterPath}
            onError={() => setImgLoadingErr()}
            onLoad={() => setImgLoadSuccess()}
          />
        </ImgTitleContainer>
      )}

      {/* On hover over Modal is activated with a transition. Extra data added. */}
      {itemHoverActive
        && (
        <Modal
          id="slider-item-modal"
          height={itemDimensions.height}
          width={itemDimensions.width}
          left={itemDimensions.left}
          top={itemDimensions.top}
          bottom={itemDimensions.bottom}
          right={itemDimensions.right}
        >
          <Transition
            appear
            in={itemHoverTransition}
            timeout={duration}
            onExited={() => handleModalDismount()}
          >
            {(state) => (
              <ModalItem
                style={{
                  ...defaultStyle,
                  ...transitionStyles[state],
                }}
                id="modal-item"
                coordsLeft={itemDimensions.left}
                coordsTop={itemDimensions.top}
                itemWidth={itemDimensions.width}
                onMouseLeave={() => handleMouseOut()}
                ref={divRef}
              >
                {/* All things that are mounted and included in transition animation */}
                <ImgTitleContainer>
                  <MediaTitle>
                    {sliderItemData.mediaTitle}
                  </MediaTitle>
                  <SliderItemImage
                    alt="Slider image"
                    src={sliderItemData.posterPath}
                    onError={() => setImgLoadingErr()}
                    onLoad={() => setImgLoadSuccess()}
                    mouseOver
                  />
                </ImgTitleContainer>
                <ItemDetails>
                  <ButtonsContainer id="buttons">
                    <Link
                      key="watch"
                      to="/watch"
                    >
                      <RoundPlayButton />
                    </Link>
                    <Link
                      key="epsInfoBox"
                      to={{
                        pathname: `${location.pathname}/epsinfobox/${mediaType}/${mediaId}`,
                        state: { background: location },
                      }}
                    >
                      <RoundEpsAndInfoButton onClick={handleEpsAndInfoButtonClick} />
                    </Link>
                  </ButtonsContainer>
                  <InfoContainer id="media-info">
                    {sliderItemData.ageRating
                      && (
                      <AgeRatingContainer>
                        {sliderItemData.ageRating}
                      </AgeRatingContainer>
                      )}

                    {sliderItemData.ageRating
                      && (
                      <div>
                        &nbsp;
                        &nbsp;
                      </div>
                      )}

                    {mediaType === 'tv' && sliderItemData.runtimeOrNumberOfSeasons.length > 1 && `${sliderItemData.runtimeOrNumberOfSeasons.length} Seasons`}
                    {mediaType === 'tv' && sliderItemData.runtimeOrNumberOfSeasons.length === 1 && `${sliderItemData.runtimeOrNumberOfSeasons.length} Season`}
                    {mediaType === 'movie' && sliderItemData.runtimeOrNumberOfSeasons}
                  </InfoContainer>
                  <GenreContainer id="genres">
                    {sliderItemData.genres[0]}
                    {sliderItemData.genres[1] && (
                      <div>
                        &nbsp;
                        Â·
                        &nbsp;
                        {sliderItemData.genres[1]}
                      </div>
                    )}
                  </GenreContainer>
                </ItemDetails>
              </ModalItem>
            )}
          </Transition>
        </Modal>
        )}
    </ItemContainer>
  );
};

export default SliderItem;
