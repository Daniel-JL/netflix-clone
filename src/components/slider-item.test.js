import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  MemoryRouter,
  Router,
} from 'react-router-dom';
import fetch from "jest-fetch-mock";
import '@testing-library/jest-dom/extend-expect';

import { createMemoryHistory } from 'history';
import { SliderItem } from './slider-item';

beforeEach(() => {
  fetch.resetMocks();
});
// jest.mock('node-fetch');
// global.fetch=jest.fn();

describe('Slider item', () => {
  it('should render correctly', () => {
    const wrapper = renderer.create(<SliderItem />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch data correctly', async () => {
    fetch.mockResponseOnce(JSON.stringify({ data: { backdrop_path: '/mGVrXeIjyecj6TKmwPVpHlscEmw.jpg' }}));
    const history = createMemoryHistory();
    const { findByAltText } = render(
      <Router history={history}>
        <SliderItem />
      </Router>,
    );

    const element1 = await findByAltText('Slider image');
    expect(element1).toBeInTheDocument();
  });

  it('should display image as long as there is no image error', async () => {
    fetch.mockResponseOnce(JSON.stringify({ data: { backdrop_path: '/mGVrXeIjyecj6TKmwPVpHlscEmw.jpg' }}));
    const history = createMemoryHistory();
    const { findByAltText } = render(
      <Router history={history}>
        <SliderItem />
      </Router>,
    );

    const element1 = await findByAltText('Slider image');
    expect(element1).toBeInTheDocument();
  });

  it('should play trailer video if available on mouseover', () => {

  });

  it('should play movie/show if play button is clicked', () => {

  });

  it('should add movie to My List if + button is pressed', () => {

  });

  it('should highlight thumb symbol if movie is liked or disliked', () => {

  });

  it('should open episodes and info box if More Info button is pressed', () => {

  });

  it('should show descriptive tooltip when mouseover any button except play', () => {

  });
});
