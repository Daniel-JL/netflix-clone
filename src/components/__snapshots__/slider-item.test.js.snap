// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider item should render correctly 1`] = `
<div
  className="sc-AxjAm jrPXsu"
/>
`;
